### PR TITLE
Feature/podaac 2552

### PR DESCRIPTION
**Pull request by jocelynmallon**
Intercept general exceptions being thrown while parsing the CNM messages, and ensure we send SNS failure messages, before re-throwing the exception.

### CODE REVIEW

File: src/main/java/gov/nasa/cumulus/CNMResponse.java [hunk 1]
```diff
@@ -58,19 +58,30 @@ public void handleRequestStreams(InputStream inputStream, OutputStream outputStr
 	public static JsonObject getResponseObject(String exception) {
 		JsonObject response = new JsonObject();
 
-		if(exception == null || new String("").equals(exception) ||  new String("None").equals(exception) ||  new String("\"None\"").equals(exception)){
-			//success
+        if (exception == null || exception.isEmpty() || exception.equals("None") || exception.equals("\"None\"")) {
```
Comment (yenes56) [inline src/main/java/gov/nasa/cumulus/CNMResponse.java:61]: use apache lang3 StringUtils class which is null safe
it has trim(), equals ... static functions

File: src/main/java/gov/nasa/cumulus/CNMResponse.java [hunk 2]
```diff
@@ -58,19 +58,30 @@ public void handleRequestStreams(InputStream inputStream, OutputStream outputStr
...
+
+            // PODAAC-2552 - handle general exceptions being thrown
+            if (exception.equalsIgnoreCase("CNMResponse Exception")) {
+                error = exception;
+                causeString = "Unknown cause, CNMResponse likely threw an exception";
+            } else {
+                workflowException = new JsonParser().parse(exception).getAsJsonObject();
+                error = workflowException.get("Error").getAsString();
+                causeString = workflowException.get("Cause").getAsString();
+            }
```
Comment (yenes56) [inline src/main/java/gov/nasa/cumulus/CNMResponse.java:83]: Can you please reformat the code?   The original one was all indented correctly.  You can easily do that if you are using intelliJ IDE

File: src/main/java/gov/nasa/cumulus/CNMResponse.java [hunk 3]
```diff
@@ -182,6 +192,32 @@ public String getError(JsonObject input, String key){
 		return exception;
 	}
 
+	public static String generateGeneralError(String cnm) {
+        JsonElement jelement = new JsonParser().parse(cnm);
```
Comment (ngachung) [inline src/main/java/gov/nasa/cumulus/CNMResponse.java:196]: When the original CNM is invalid, such as when it is missing product, the cnmtogranule step of the workflow fails, so the OriginalCNM is actually null which will cause this line here to fail with IllegalStateException. In this case SNS message never gets sent. In the case where cnmtogranule step fails, the input to cnmresponse is the original CNM. 

Sample input:

```
{
  "input": {
    "collection": "L0A_KCAL_Packet",
    "version": "1.0",
    "provider": "PODAAC",
    "submissionTime": "2020-07-05T20:42:32.682418",
    "identifier": "45150e51-955a-11ea-80e5-acde48001122"
  },
  "config": {
    "OriginalCNM": null,
    "type": "sns",
    "response-endpoint": "arn:aws:sns:us-west-2:111111111111:podaac-sndbx-cumulus-provider-response-sns",
    "region": "us-west-2",
    "WorkflowException": {
      "Error": "cumulus_message_adapter.message_parser.MessageAdapterException",
      "Cause": "{\\\"errorMessage\\\":\\\"An error occurred in the Cumulus Message Adapter: java.lang.NullPointerException\\\",\\\"errorType\\\":\\\"cumulus_message_adapter.message_parser.MessageAdapterException\\\",\\\"stackTrace\\\":[\\\"cumulus_message_adapter.message_parser.MessageParser.HandleMessage(MessageParser.java:100)\\\",\\\"cumulus_message_adapter.message_parser.MessageParser.RunCumulusTask(MessageParser.java:117)\\\",\\\"gov.nasa.cumulus.CnmToGranuleHandler.handleRequestStreams(CnmToGranuleHandler.java:55)\\\",\\\"sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\\\",\\\"sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\\\",\\\"sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\\\",\\\"java.lang.reflect.Method.invoke(Method.java:498)\\\"],\\\"cause\\\":{\\\"errorMessage\\\":\\\"java.lang.NullPointerException\\\",\\\"errorType\\\":\\\"java.lang.NullPointerException\\\",\\\"stackTrace\\\":[\\\"gov.nasa.cumulus.CnmToGranuleHandler.PerformFunction(CnmToGranuleHandler.java:126)\\\",\\\"cumulus_message_adapter.message_parser.MessageParser.HandleMessage(MessageParser.java:76)\\\",\\\"cumulus_message_adapter.message_parser.MessageParser.RunCumulusTask(MessageParser.java:117)\\\",\\\"gov.nasa.cumulus.CnmToGranuleHandler.handleRequestStreams(CnmToGranuleHandler.java:55)\\\",\\\"sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\\\",\\\"sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\\\",\\\"sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\\\",\\\"java.lang.reflect.Method.invoke(Method.java:498)\\\"]}}"
    }
  },
  "messageConfig": {
    "outputs": [
      {
        "source": "{$.cnm}",
        "destination": "{$.meta.cnmResponse}"
      },
      {
        "source": "{$.input.input}",
        "destination": "{$.payload}"
      }
    ]
  },
  "cumulus_config": {
    "state_machine": "arn:aws:states:us-west-2:111111111111:stateMachine:podaac-sndbx-cumulus-IngestWorkflow",
    "execution_name": "d0473918-e64c-4bf5-82fc-5c4e4bcb3e2b"
  }
}
```

**Review by ngachung — CHANGES_REQUESTED** [review]

File: src/main/java/gov/nasa/cumulus/CNMResponse.java (hunk 3 above)
Comment (yenes56) [reply src/main/java/gov/nasa/cumulus/CNMResponse.java:196]: Yes, after some integration testing (by making CNMToGranule throws exception purposely), the CNM object passed into this function is actually null or maybe empty string. My integration testing actually causes NullpointerException thrown on line #197(so I suspect it is an empty string but not 100% sure).  However, in terms of implementation, I think StringUtils.isEmpty(cnm) will be sufficient to verify the case

if it is empty, there should be no need to do :

        JsonElement jelement = new JsonParser().parse(cnm);
        JsonObject inputKey = jelement.getAsJsonObject();

**Comment by jocelynmallon** [comment]
Closing this PR to rebase and clean up the changes.

**Pull request closed without merging**